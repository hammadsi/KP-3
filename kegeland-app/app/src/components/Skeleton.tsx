import React, {useEffect, useRef} from 'react';
import {Animated, Easing, View, ViewStyle} from 'react-native';
import {useTheme} from 'react-native-paper';

export type SkeletonProps = {
  mode?: 'rounded' | 'square';
  width?: number;
  height?: number;
  style?: ViewStyle;
  children?: React.ReactNode;
};

/**
 * Component for rendering a component skeleton while a component is rendering.
 * @param props the props
 * @see {@link SkeletonProps}
 */
const Skeleton: React.FC<SkeletonProps> = (props) => {
  const {colors} = useTheme();
  const opacity = useRef(new Animated.Value(0.3));

  const style: ViewStyle = {
    width: props.width,
    height: props.height,
    borderRadius: props.mode === 'rounded' && props.width ? props.width / 2 : 0,
  };

  /**
   * Initiate animation
   */
  useEffect(() => {
    Animated.loop(
      Animated.sequence([
        Animated.timing(opacity.current, {
          toValue: 0.1,
          useNativeDriver: true,
          easing: Easing.cubic,
          duration: 0,
        }),
        Animated.timing(opacity.current, {
          toValue: 0.3,
          useNativeDriver: true,
          duration: 1000,
        }),
        Animated.timing(opacity.current, {
          toValue: 0.1,
          useNativeDriver: true,
          duration: 2000,
        }),
      ]),
    ).start();
  }, [opacity]);
  return (
    <Animated.View
      style={[
        {
          opacity: opacity.current,
          backgroundColor: colors.onSurface,
        },
        style,
        props.style,
      ]}>
      <View style={{}} />
      {props.children}
    </Animated.View>
  );
};

Skeleton.defaultProps = {
  mode: 'square',
};

export default Skeleton;
