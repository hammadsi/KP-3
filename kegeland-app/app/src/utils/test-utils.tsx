/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
