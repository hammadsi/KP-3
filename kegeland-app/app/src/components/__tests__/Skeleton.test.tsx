import React from 'react';

import {render} from '~utils/test-utils';

import Skeleton from '../Skeleton';

describe('Test Skeleton-component', () => {
  it('should render correctly', () => {
    const component = <Skeleton />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
