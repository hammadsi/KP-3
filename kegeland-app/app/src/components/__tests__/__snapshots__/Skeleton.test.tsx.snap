// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Skeleton-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#000000",
        "borderRadius": 0,
        "height": undefined,
        "opacity": 0.1,
        "width": undefined,
      }
    }
  >
    <View
      style={Object {}}
    />
  </View>
</View>
`;
